src/nrniv/neuronapi.cpp

```diff
@@ -158,159 +168,277 @@ void nrn_segment_diam_set(Section* const sec, const double x, const double diam)
+double nrn_rangevar_get(Section* sec, double x, const char* var_name) {
+    Symbol* sym = hoc_table_lookup(var_name, hoc_built_in_symlist);
     return *nrn_rangepointer(sec, sym, x);
 }
 
-void nrn_rangevar_set(Symbol* sym, Section* sec, double x, double value) {
+void nrn_rangevar_set(Section* sec, double x, const char* var_name, double value) {
+    Symbol* sym = hoc_table_lookup(var_name, hoc_built_in_symlist);
     *nrn_rangepointer(sec, sym, x) = value;
 }
 
-void nrn_rangevar_push(Symbol* sym, Section* sec, double x) {
-    hoc_push(nrn_rangepointer(sec, sym, x));
+RangeVar nrn_rangevar_new(Section* sec, double x, const char* var_name) {
+    auto* node = static_cast<Segment*>(node_exact(sec, x));
+    if (strcmp(var_name, "v") == 0) {  // v doesnt require symbol
+        return {node, x, nullptr};
+    } else {
+        return {node, x, hoc_table_lookup(var_name, hoc_built_in_symlist)};
+    }
 }
 
-nrn_Item* nrn_allsec(void) {
-    return static_cast<nrn_Item*>(section_list);
+NrnList* nrn_allsec(void) {
+    return static_cast<NrnList*>(section_list);
 }
 
-nrn_Item* nrn_sectionlist_data(Object* obj) {
+NrnList* nrn_sectionlist_data(Object* obj) {
     // TODO: verify the obj is in fact a SectionList
-    return (nrn_Item*) obj->u.this_pointer;
+    return static_cast<NrnList*>(obj->u.this_pointer);
 }
 
 /****************************************
  * Functions, objects, and the stack
  ****************************************/
 
+static constexpr int MAX_ERR_LEN = 200;
+static char nrn_stack_error[MAX_ERR_LEN] = "";
+
+char const* nrn_class_name(const Object* obj) {
+    return obj->ctemplate->sym->name;
+}
+
+const char* nrn_stack_err() {
+    if (nrn_stack_error[0]) {
+        return nrn_stack_error;
+    } else {
+        return NULL;
+    }
+}
+
+void clear_errors() {
+    nrn_stack_error[0] = 0;
+}
+
 Symbol* nrn_symbol(char const* const name) {
     return hoc_lookup(name);
 }
 
+Symbol* nrn_method_symbol(Object* obj, char const* const name) {
+    return hoc_table_lookup(name, obj->ctemplate->symtable);
+}
+
 int nrn_symbol_type(Symbol const* sym) {
     // TODO: these types are in parse.hpp and are not the same between versions,
     // so we really should wrap
     return sym->type;
 }
 
-void nrn_symbol_push(Symbol* sym) {
-    hoc_pushpx(sym->u.pval);
+Object* nrn_object_pop() {
+    // NOTE: the returned object should be unref'd when no longer needed
+    Object** obptr;
+    try {
+        obptr = hoc_objpop();
+    } catch (const std::exception& e) {
+        snprintf(nrn_stack_error, MAX_ERR_LEN, "nrn_object_pop: %s", e.what());
+        return nullptr;
+    }
+    Object* new_ob_ptr = *obptr;
+    hoc_obj_ref(new_ob_ptr);
+    hoc_tobj_unref(obptr);
+    return new_ob_ptr;
 }
 
-void nrn_double_push(double val) {
-    hoc_pushx(val);
+static int stack_push_args(const char arg_types[], va_list args) {
+    for (int n_arg = 0; n_arg < 32; ++n_arg) {  // max 32 args
```

Comment:
why a limit of 32? Is that intrinsic to HOC as configured? If so, can we have both this and that use the same constant instead of a magic number here?

Reply:
If there's a related hoc limit defined in a constant I'd be happy to use.